### DISCUSSION TITLE
Elevator Encoder Options

### DISCUSSION THREAD

**Discussion by PChild** [team401/2025-Robot-Code#2] [Q&A] [answered]
Do you all have preferences about how we handle an encoder for elevator position?

The choices as I see them:

- Use only the internal encoder in one of the Krakens
     - Assume that the system is fully down when code starts, 
     - Have a current or velocity homing routine that zeros the system at bottom of motion. 
     - Use motion profiling on the Krakens. 
     - The Krakens are behind a gearing stage or two so this will have some backlash.
     
- Through bore encoder on the elevator spool shaft
     - Use this on code boot to seed the position of the Kraken internal encoder. 
     - Still needs a homing routine
     - Still has backlash
     
- CANcoder on the elevator spool shaft
     - Still need homing routine
     - Can be used as fused CANcoder to eliminate backlash
     
- Two CANcoders set up for Chinese Remainder Theorem
     - Backlash should be very low if any
     -  No homing routine needed - CRT code should seed position.
     - Can use fused CANcoder with one of the CANcoders for ongoing control of the system.

**Comment by PChild** [comment]
@jkleiber @aidnem Thoughts? My initial choice would be to do CRT but I just think it's cool.

**Reply by aidnem** [reply]
@PChild do CANCoders wrap around after 1 rotation? I'm trying to figure out if we could use PID and motion magic on the krakens with CRT. If we can't I think that would be the first downside of CRT that comes to mind (other than being harder/more annoying to program), so I'd be fine with doing CRT.

That's all I thought of at first glance but I'll think about it more.

**Reply by jkleiber** [reply]
I think the idea is that if the CANCoders cannot go more than 1 rotation if we are doing chinese remainder theorem. So it would be a simple matter of calibrating the CANCoders to be at 0 for the bottom position and then use CRT to compute the displacement of the elevator off of the bottom.

I don't think we'd be able to use motion magic on the krakens with CRT, but I think doing motion profiling on the roboRIO is worth it if we get a measurement system with no backlash.

**Reply by jkleiber** [reply]
Wait, but thinking about this more I think the encoders wrapping isn't an issue since they are effectively the "remainders" in the theorem. So as long as we can get output in absolute ticks and know the reductions we should be fine as long as we get an offset?

Pros of CRT:
- No homing method required (having a homing step is a system limitation that we should avoid imo)
- Limited backlash

Cons of CRT:
- Additional cost (not much though)
- Additional code complexity

The pros are worth the cons imo

**Reply by PChild** [reply]
Yes, that matches my thinking. CRT should be no homing, very limited backlash, and let us still use motion magic.

**Reply by jkleiber** [reply]
I'm don't think embedded motion magic is on the table with CRT? 

It would be one actuator with 2 sensors, and we would be computing the true position onboard the roborio. We could do motionmagic/wpilib trajectory stuff there though if that's what you mean @PChild

**Reply by PChild** [reply]
Maybe I'm thinking about it wrong but my idea was that we use CRT and compute the position of the system only on code init to seed the position of the Kraken. We then use just the first encoder in the system as a remote feedback device with the Kraken so we see low / zero backlash instead of 1-3 gearing stages worth from the MaxPlanetary.

**Reply by jkleiber** [reply]
But assuming the encoder is going to make multiple rotations over the range of the elevator's motion, I don't think we would have a valid control target for motion magic (at least in absolute mode)

Maybe it would be possible to control with a relative position, but then we'll need to maintain the offset in code and likely do a bunch of calibration (number of ticks of encoder 1 vs displacement sounds like a bad time)

**Reply by PChild** [reply]
Yes, we'd have to use the relative position of the encoder but that's the case in all of the options I proposed except for CRT with RIO side control. There's a `Feedback.SensorToMechanismRatio`  config option that we can set on the Kraken that should eliminate calibration.

**Reply by aidnem** [reply]
I wonder if through combining sensor to mechanism ratio and the rotor to mechanism ratio we could control the motors through multiple rotations of the CANCoder without it losing its position? My initial comment about ruling out embedded motion magic was that we wouldn't be able to control it to an arbitrary position if it thought it could only wrap around once.
I assume this is what you're saying as well @PChild

**Reply by PChild** [reply, marked as answer]
You can control a mechanism through multiple rotations of a CANcoder. You can access both [position](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/hardware/core/CoreCANcoder.html#getPosition()) which is limited to +/- 16384 rotations, and [absolute position](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/hardware/core/CoreCANcoder.html#getAbsolutePosition()) which is either  [-0.5, 0.5) or [0, 1) rotations depending on your `AbsoluteSensorRange`. 

We would use the absolute position data on the RIO for CRT. Normal elevator control would then use just the position data.

**Reply by jkleiber** [reply]
Oh yeah so we basically can see the absolute position in CRT but then calculate how many ticks it'll take from the current position to get to the goal (delta_ticks) and then do MM to current_ticks + delta_ticks = goal_ticks

**Reply by aidnem** [reply]
We don't even need to use the position delta for control. Phoenix 6 lets you [set the position](https://api.ctr-electronics.com/phoenix6/latest/java/com/ctre/phoenix6/hardware/core/CoreCANcoder.html#setPosition(edu.wpi.first.units.measure.Angle)) of a CANCoder, so after calculating the position using CRT we can just set the positions of the encoders to reflect the actual rotations from zero.
Then, to control to a height we just convert height -> spool rotations -> CANCoder rotations and then tell Phoenix 6 closed loop to control to the goal CANCoder rotations.
